### PR TITLE
nixos/systemd: provide a typed interface to Exec directives

### PR DESCRIPTION
**Pull request by pennae**
###### Motivation for this change

implements #154123. only very lightly tested so far. documentation also still needs updating/writing.

@roberth @NixOS/systemd 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/systemd-unit-options.nix [hunk 1]
```diff
@@ -309,6 +430,16 @@ in rec {
...
+        Mutually exclusive with stopCommands.
+      '';
+    };
+
+    postStopCommands = mkOption {
+      type = types.listOf serviceCommandType;
+      default = [ ];
+      description = ''
+        Commands executed after the service's main process
+        has exited.Mutually exclusive with postStop.
```
Comment (roberth) [inline nixos/lib/systemd-unit-options.nix:442]:
```suggestion
        has exited. Mutually exclusive with postStop.
```

File: nixos/lib/utils.nix [hunk 2]
```diff
@@ -45,6 +45,23 @@ rec {
...
+  # substitution for the directive.
+  escapeSystemdExecArg = arg:
+    let
+      s = builtins.toJSON (if builtins.isPath arg then "${arg}" else toString arg);
+    in
+      replaceChars [ "%" "$" ] [ "%%" "$$" ] s;
+
+  # Quotes a list of arguments into a single string for use in a Exec*
+  # line.
+  escapeSystemdExecArgs = concatMapStringsSep " " escapeSystemdExecArg;
```
Comment (roberth) [inline nixos/lib/utils.nix:63]: I don't think we need this anymore.

File: nixos/lib/utils.nix [hunk 3]
```diff
@@ -45,6 +45,23 @@ rec {
...
    (removePrefix "/" s);
 
+  # Quotes an argument for use in Exec* service lines.
+  # systemd accepts "-quoted strings with escape sequences, toJSON produces
+  # a subset of these.
+  # Additionally we escape % to disallow expansion of % specifiers. Any lone ;
+  # in the input will be turned it ";" and thus lose its special meaning.
+  # Every $ is escaped to $$, this makes it unnecessary to disable environment
+  # substitution for the directive.
+  escapeSystemdExecArg = arg:
```
Comment (roberth) [inline nixos/lib/utils.nix:55]: You could move this to the systemd module. `utils.nix` is not a great place to be.

File: nixos/lib/systemd-unit-options.nix [hunk 4]
```diff
@@ -231,6 +231,79 @@ in rec {
...
+        type = types.bool;
+        description = ''
+          Do not consider it a failure when this command returns a non-zero
+          exit status.
+        '';
+      };
+
+      unrestricted = mkOption {
+        default = null;
+        type = with types; nullOr (enum [ "all" "credentials" ]);
```
Comment (roberth) [inline nixos/lib/systemd-unit-options.nix:262]:
```suggestion
        type = types.enum [ null "all" "credentials" ]);
```

File: nixos/lib/systemd-unit-options.nix [hunk 5]
```diff
@@ -231,6 +231,79 @@ in rec {
...
+      args = mkOption {
+        default = [ ];
+        type = with types; listOf (oneOf [ str systemdCommandExpansion int float ]);
+        description = ''
+          Arguments passed to the command. Arguments will not be expanded
+          by systemd unless requested.
+
+          To expand an environment variable `X` use `{ env = "X"; }`.
+
+          To apply all systemd substitutions to a string use `{ substitute = "..."; }`.
```
Comment (roberth) [inline nixos/lib/systemd-unit-options.nix:296]: Reference the "specifiers"

```suggestion
          To apply all systemd substitutions to a string use <literal>{ substitute = "..."; }</literal>.
          
          See "Specifiers" in <literal>man systemd.unit</literal> for `%` syntax.
```

and yeah, it's still docbook here. :/

File: nixos/lib/systemd-unit-options.nix [hunk 6]
```diff
@@ -231,6 +231,79 @@ in rec {
...
+        '';
+      };
+    };
+
+    config = {
+      exe = mkIf (config.script != null)
+        (mkDerivedConfig (options.script) builtins.toFile);
```
Comment (roberth) [inline nixos/lib/systemd-unit-options.nix:303]: You'll need a `writeScriptBin` here, both to make it executable and to allow store path references.

File: nixos/tests/systemd-escaping.nix [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  args = [ "a%Nything" "lang=\${LANG}" ";" "/bin/sh -c date" ./empty-file ];
+in
+{
+  name = "systemd-escaping";
+
+  machine = { pkgs, lib, utils, ... }: {
+    systemd.services.echo =
+      { description = "Echo to the journal";
+        serviceConfig.Type = "oneshot";
+        serviceConfig.ExecStart = ''
```
[the remark this answers is not in the document]
Comment (pennae) [reply nixos/tests/systemd-escaping.nix:21]: we'll replace this with a test for the new `*Commands` attrs once the interface can be considered settled. right now it's just here to keep escapeSystemdExecArg honest :)

File: nixos/lib/systemd-unit-options.nix (hunk 5 above)
Comment (pennae) [reply nixos/lib/systemd-unit-options.nix:296]: also added a reference to environment substitution because that can happen as well.

**Comment by aanderse** [comment]
Some abstractions are good. Some aren't. I think `script`, `preStart`, `postStart`, and `environment` are pretty simple to understand and almost directly map to `systemd` declarations - because they don't do much. These declarations also keep us pretty close to upstream units, to such a degree that we can often just add `systemd.packages = [ pkgs.foo ];` and be done with it.

When we start adding complex abstractions we make ourselves even more alien to the rest of the gnu/linux ecosystem and their developers. We increase the already incredibly steep learning curve with ecosystem specific knowledge that isn't transferable at all.

This is ok, but we have to ask ourselves "is it worth it?" every time we do. So I ask: is this abstraction worth it? Or are we better off taking another route?

**Comment by pennae** [comment]
an alternative solution is to add an analog of `escapeShellArgs` for systemd exec directives (#154113). we don't care much which one ends up being chosen, as long as there's a war for people to avoid making the same mistake over and over again. :)

**Comment by roberth** [comment]
I'd like to differentiate this feature from abstractions as those are commonly talked about. Whereas the feature is an abstraction in the sense of creating an indirection, it is not an abstraction in the sense of eliding "irrelevant" aspects, or encapsulating details. Much of the (correct) criticism of abstraction revolves around the latter parts.

> Some abstractions are good. Some aren't.

Yes, and even at the same time. NixOS is great, because it doesn't force you into either one.
For example, you can enable our "abstracted" services, like `services.postgresql`, or define your own, using `systemd.services`.

Similarly, this added interface doesn't require you to avoid `serviceConfig`. You can still define services that way and stay visibly close to an upstream unit file.

On the other hand, when that is not a requirement, you can make use of well-defined, predictable options with consistent searchable documentation.

Consider a backend team developing custom company services on a Nix stack. Hopefully they've picked up on actual DevOps, so they'll have someone on the team who's familiar at least with the relevance of those symbols, etc, but the others should be able to interact easily with the config file as well. For them, a cryptic DSL like systemd's Exec lines only gets in the way.

> When we start adding complex abstractions we make ourselves even more alien to the rest of the gnu/linux ecosystem and their developers.

Good thing this one isn't complex. Not even much of an abstraction; just a sensible format for defining service commands.
I find it hard to imagine that someone would like to express important security properties using otherwise meaningless symbols like `+` or `!`, so it would seem that of instead of alienation, we should see appreciation.

> So I ask: is this abstraction worth it?

In my view, absolutely.

I appreciate you keeping us alert to this.

**Comment by RaitoBezarius** [comment]
I'm planning to adopt, with @pennae permission, this PR and push it to the finish line in the next days/weeks.

**Comment by mweinelt** [comment]
> > I'm planning to adopt, with @pennae permission, this PR and push it to the finish line in the next days/weeks.
> 
> absolutely!

Did anything come off of this?

**Comment by RaitoBezarius** [comment]
Still on my radar but very deprioritized. I don't have any working tree, if anyone is interested, feel free to go on it before me.

**Pull request closed without merging**